main.go

```diff
@@ -91,26 +102,56 @@ func main() {
 			tlsConn, err := vhost.TLS(c)
 			if err != nil {
 				log.Printf("Error accepting new connection - %v", err)
+				return
 			}
+			defer tlsConn.Close()
+
 			if tlsConn.Host() == "" {
 				log.Printf("Cannot support non-SNI enabled clients")
 				return
 			}
-			connectReq := &http.Request{
-				Method: "CONNECT",
-				URL: &url.URL{
-					Opaque: tlsConn.Host(),
-					Host:   net.JoinHostPort(tlsConn.Host(), "443"),
-				},
-				Host:   tlsConn.Host(),
-				Header: make(http.Header),
+
+			remoteHost := net.JoinHostPort(tlsConn.Host(), "443")
+			conn, err := dialer.Dial("tcp", remoteHost)
+			if err != nil {
+				log.Printf("Error establishing connection to %s: %v", remoteHost, err)
+				return
 			}
-			resp := dumbResponseWriter{tlsConn}
-			proxyServer.ServeHTTP(resp, connectReq)
+			defer conn.Close()
+
+			if proxyConnection, ok := conn.(*proxy.Connection); ok {
+				clientHost, _, err := net.SplitHostPort(c.RemoteAddr().String())
+				if err != nil {
+					log.Printf("Failed to get host from address %s: %v", c.RemoteAddr(), err)
+					return
+				}
+				if err := proxyConnection.ConnectTo(conn, remoteHost, sm.Hash(clientHost)); err != nil {
+					log.Printf("Error establishing CONNECT tunnel to %s: %v", remoteHost, err)
+					return
+				}
+			}
+
+			copyAndWait(conn, tlsConn)
 		}(c)
 	}
 }
 
+func copyAndWait(src, dst io.ReadWriter) {
+	var wg sync.WaitGroup
+	wg.Add(2)
+
+	go func() {
+		io.Copy(dst, src)
```

Comment:
No, we didn't test it. 
But `io.Copy` is probably the fastest way that we can use in golang for connecting streams. 
I'd say it will require more resources, but definitely performance will not drop 2x.